File: tests/fibonacci_Dict.jl
#!/usr/bin/env julia
const dict = Dict{Int,BigInt}(0=>BigInt(0), 1=>BigInt(1))

function fib(n::Int)::Any
  get!(dict, n) do
    m::Int = n >> 1
    iseven(n) ? fib(m)*(fib(m-1) << 1 + fib(m)) : fib(m+1)^2 + fib(m)^2
  end
end

function main()
  line = readline()
  n::Int = tryparse(Int, line)
  n < 0 && error("The input value must be a positive integer")
  println(fib(n))
end

main()

File: tests/fibonacci_Dict.py
#!/usr/bin/env python3
dic = {0:0, 1:1}
def fib(n):
  if n in dic:
      return dic[n]

  m = n >> 1
  if (n % 2 == 0):
      dic[n]=fib(m)*((fib(m-1) << 1)+fib(m))
  else:
      dic[n]=fib(m+1)**2 + fib(m)**2
  return dic[n]

line = input()
try:
    n = int(line)
except ValueError:
    n = -1

if (n < 0):
    print("The input value must be a positive integer")
else:
    print(fib(n))
